File: August 2022/August_11.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        isBST = true;
        helper(root);
        return isBST;
    }
    
    boolean isBST;
    
    private int[] helper(TreeNode root){
        int[] arr = new int[2];
        if(root.left == null)
            arr[0] = root.val;
        else{
            int[] left = helper(root.left);
            if(left[1] >= root.val)
                isBST = false;
            arr[0] = left[0];
        }
        if(root.right == null)
            arr[1] = root.val;
        else{
            int[] right = helper(root.right);
            if(right[0] <= root.val)
                isBST = false;
            arr[1] = right[1];
        }
        return arr;
    }
}

File: August 2022/August_11.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBST;
    
    pair<int, int> helper(TreeNode* root){
        pair<int, int> p;
        if(root->left == nullptr)
            p.first = root->val;
        else{
            pair<int, int> left = helper(root->left);
            if(left.second >= root->val)
                isBST = false;
            p.first = left.first;
        }
        if(root->right == nullptr)
            p.second = root->val;
        else{
            pair<int, int> right = helper(root->right);
            if(right.first <= root->val)
                isBST = false;
            p.second = right.second;
        }
        return p;
    }
    
    bool isValidBST(TreeNode* root) {
        isBST = true;
        helper(root);
        return isBST;
    }
};
